pennylane/transforms/decompose.py

```diff
@@ -901,74 +895,124 @@ def _operator_decomposition_gen(  # pylint: disable=too-many-arguments,too-many-
+    # For compatibility reasons, we don't raise an error when graph mode is not enabled.
+    # We simply disregard the weights and treat the dictionary as just a set of gates.
+    if not enabled_graph():
+        gate_set = set(gate_set.keys())
+        warnings.warn(
+            "Gate weights were provided to a non-graph-based decomposition. These will be ignored."
+        )
+    return gate_set
+
+
+@_process_gate_set.register
+def _(gate_set: Iterable):
+    return set(gate_set)
+
+
+@singledispatch
+def _process_gate_set_contains(gate_set):  # pylint: disable=unused-argument
+    raise TypeError("Invalid gate_set type. Must be an iterable, dictionary, or function.")
+
+
+@_process_gate_set_contains.register
+def _(gate_set: Iterable):
+    # The gate_set could be a mix of operator names and operator types. We need to wrap this
+    # in a gate_set_contains function that checks if either the name of the operator is within
+    # the names in the gate set, or if the type of the operator is within the types.
+    gate_types = tuple(gate for gate in gate_set if isinstance(gate, type))
+    gate_names = {translate_op_alias(gate) for gate in gate_set if isinstance(gate, str)}
+
+    def gate_set_contains(op: Operator) -> bool:
+        return (op.name in gate_names) or isinstance(op, gate_types)
+
+    return gate_set, gate_set_contains
+
+
+@_process_gate_set_contains.register
+def _(gate_set: NoneType):  # pylint: disable=unused-argument
+    # At the beginning of the function we already handled the special case for when neither
+    # gate_set nor stopping_condition is provided. Here we handle the case when gate_set
+    # is not provided but stopping_condition is. This would only be valid with graph disabled.
+    gate_set = set()
+
+    # pylint: disable=unused-argument
+    def gate_set_contains(op: Operator) -> bool:
+        return False
+
+    if enabled_graph():
         raise TypeError(
-            "Specifying both gate_set and stopping_condition is only supported with the new "
-            "experimental graph-based decomposition system. Use qml.decomposition.enable_graph() "
-            "to enable the new system."
+            "The gate_set argument is required when the graph-based decomposition system "
+            "is enabled via qml.decomposition.enable_graph()"
         )
 
-    if gate_set is None:
-        gate_set = set(ops.__all__)
+    return gate_set, gate_set_contains
 
-    if isinstance(gate_set, (str, type)):
-        gate_set = {gate_set}
 
-    if isinstance(gate_set, dict):
+@_process_gate_set_contains.register
+def _(gate_set: Callable):
+    # This branch exists for backwards compatibility reasons. I forgot to bring this up
+    # as a v0.44 deprecation but maybe we should deprecate providing a function to the
+    # gate_set argument because only the stopping_condition is supposed to be a function.
+    # The deprecation is proposed in sc-102183
```

Comment:
```suggestion
    # This branch exists for backwards compatibility reasons. It was overlooked
    # as a v0.44 deprecation. It is suggested that providing a function to the
    # gate_set argument should be deprecated, because only the stopping_condition is supposed to be a function.
    # The deprecation is proposed in sc-102183
```